### PR TITLE
[SPARK-7286] [SQL] Deprecate !== in favour of =!=

### PR DESCRIPTION
**Pull request by jodersky**
Fixes subtle issues related to operator precedence, as discussed in SPARK-7286.

I'm not entirely sure this is the right thing to do, personally I think operator overloading is a slippery slope to symbol soup. Please feel free to comment on alternatives.

I anticipated this for Spark 1.7, let me know if I should change the deprecation warnings.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala [hunk 1]
```diff
@@ -83,7 +83,10 @@ package object dsl {
     def >= (other: Expression): Predicate = GreaterThanOrEqual(expr, other)
     def === (other: Expression): Predicate = EqualTo(expr, other)
     def <=> (other: Expression): Predicate = EqualNullSafe(expr, other)
```
Comment (jodersky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: On a side note, what does <=> do? I find it very cryptic

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: It's null safe equality test, i.e. comparing null and null should return true rather than null.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala [hunk 2]
```diff
@@ -83,7 +83,10 @@ package object dsl {
...
-    def !== (other: Expression): Predicate = Not(EqualTo(expr, other))
+    def =!= (other: Expression): Predicate = Not(EqualTo(expr, other))
+
+    @deprecated("!== does not have the same precedence as ===, use =!= instead", "1.7.0")
+    def !== (other: Expression): Predicate = this =!= other
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: this is an internal package - so it doesn't make sense to deprecate anything here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 1 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: @jodersky Hi Jakob, <=> is a null safe equal operator.  Here is a link that describes the semantics.
http://dev.mysql.com/doc/refman/5.0/en/comparison-operators.html#operator_equal-to

Comment (jodersky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: Ok, so it seems it is a well-established operator. Unfortunately <=> it also has a slightly lower precedence than =!= and ===, but since it is something known and I see no suitable alternative, I guess its best to live with it

**Comment by hvanhovell** [comment]
Why not use the more common ```<>``` symbol instead ```=!=```?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 2 above)
Comment (jodersky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: Oh, I missed that. Are you suggesting I should change without deprecation or leave the oerator name as-is?

**Comment by jodersky** [comment]
It has the same issue as !==

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 2 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: Just remove it.

Comment (jodersky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala]: @rxin should be fixed now

**Comment by jodersky** [comment]
@rxin, just wanted to check if this PR is acceptable?

**Comment by rxin** [comment]
=!= is just really ugly ... let's sleep on this and decide later. There is no rush since we are not doing it for 1.6.

**Comment by jodersky** [comment]
I agree that its not pretty, however the only other fix I see is to remove "$" for columns instead

**Comment by jodersky** [comment]
Since the next version of Spark is probably going to be a major release, would now not be a good time to consider renaming the operator as to achieve correct behavior?

**Comment by rxin** [comment]
We should definitely consider this for Spark 2.0.

**Comment by jodersky** [comment]
Is this still under consideration? I have no problem closing the issue if you think it's not a good enough solution.

**Comment by rxin** [comment]
I think we should still consider it -- just haven't had time to think about it more yet.

Can we get somebody from the scala team to also chime in here? 

cc @dragos

**Comment by dragos** [comment]
I like @jodersky's solution. As I mentioned on the JIRA, Slick uses the same operator, so at least there's a precedent, and some people might find it familiar.

**Comment by rxin** [comment]
@jodersky ok let's do this!

Rather than removing the old one, can you just deprecate it and add the new one?

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 3]
```diff
@@ -265,8 +265,26 @@ class Column(protected[sql] val expr: Expression) extends Logging {
    *
    * @group expr_ops
    * @since 1.3.0
-   */
-  def !== (other: Any): Column = withExpr{ Not(EqualTo(expr, lit(other).expr)) }
+    */
+  def =!= (other: Any): Column = withExpr{ Not(EqualTo(expr, lit(other).expr)) }
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: need to update the example in the documentation above

**Comment by rxin** [comment]
nvm you've already deprecated it. 

can you fix the doc?

**Comment by jodersky** [comment]
jenkins, test this please

**Comment by jodersky** [comment]
Btw, a general question, do you want me to squash commits as I ammend them to PRs?

**Comment by rxin** [comment]
I doesn't matter - the merge script automatically squashes them.

**Comment by jodersky** [comment]
Seems that some tests have added !==  in the mean-time, I'll update the references

**Comment by jodersky** [comment]
alright
jenkins, test this please

**Comment by jodersky** [comment]
jenkins, retest this please

**Comment by jodersky** [comment]
What's up with jenkins? retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by jodersky** [comment]
@JoshRosen, do you know why Jenkins isn't building this PR?

**Comment by JoshRosen** [comment]
Not sure; maybe Jenkins was down during the last test requests? Let me try again.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by jodersky** [comment]
Thanks! Do you know what was the issue?

**Comment by jodersky** [comment]
Another use of !== had managed to slip in.

**Comment by jodersky** [comment]
Jenkins, retest this please

**Comment by jodersky** [comment]
@JoshRosen, sorry to bother you again. Jenkins is ignoring my requests, could it be that I'm not white-listed anymore?

**Comment by jodersky** [comment]
Jenkins now seems to be stuck on a build :(
@rxin, should I close this PR and create a new one, would this fix the Jenkins response issues?

**Comment by rxin** [comment]
I'm not sure - but that's probably a good thing to try.

**Comment by jodersky** [comment]
Closing this PR, will reopen a new one
